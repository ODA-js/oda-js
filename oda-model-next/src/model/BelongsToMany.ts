import { Map, Record } from 'immutable';

import {
  IBelongsToMany,
  IBelongsToManyInit,
  IBelongsToManyStore,
  IRelationTransform,
} from '../interfaces/IBelongsToMany';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField, IFieldInit } from '../interfaces/IField';
import { EntityRef } from './EntityRef';
import { Field } from './Field';
import { Relation } from './Relation';
import { TransformRef, TransformField } from './utils';
import { IFieldContext } from '../contexts/IFieldContext';
import { Persistent } from './Persistent';

// tslint:disable-next-line:variable-name
export const DefaultBelongsToMany: IBelongsToManyStore = {
  name: null,
  title: null,
  description: null,
  fields: null,
  opposite: null,
  belongsToMany: null,
  using: null,
  //storage
  single: false,
  stored: false,
  embedded: false,
  // name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const BelongsToManyTransform: IRelationTransform = {
  belongsToMany: TransformRef(),
  fields: TransformField(),
  using: TransformRef(),
};

// tslint:disable-next-line:variable-name
export const BelongsToManyStorage = Record(DefaultBelongsToMany);

export class BelongsToMany
  extends Relation<IBelongsToManyInit, IBelongsToManyStore>
  implements IBelongsToMany {
  public get verb(): 'BelongsToMany' {
    return 'BelongsToMany';
  }
  public get ref(): IEntityRef {
    return this.store.get('belongsToMany', null);
  }
  public get belongsToMany(): IEntityRef {
    return this.store.get('belongsToMany', null);
  }
  public get using(): IEntityRef {
    return this.store.get('using', null);
  }

  protected transform(
    input: Partial<IBelongsToManyInit>,
  ): Partial<IBelongsToManyStore> {
    const result: IBelongsToManyStore = {} as any;
    if (input) {
      if (input instanceof Persistent) {
        input = input.toJS();
      }
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'belongsToMany') {
            result.belongsToMany = BelongsToManyTransform.belongsToMany.transform(
              input.belongsToMany,
              this,
            );
          } else if (f === 'using') {
            result.using = BelongsToManyTransform.using.transform(
              input.using,
              this,
            );
          } else if (f === 'fields') {
            result.fields = BelongsToManyTransform.fields.transform(
              input.fields,
              this,
            );
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(
    input: Record<IBelongsToManyStore> & Readonly<IBelongsToManyStore>,
  ): Partial<IBelongsToManyInit> {
    const result: IBelongsToManyInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (core[f] !== undefined && core[f] !== null) {
            if (f === 'belongsToMany' || f === 'ref') {
              result.belongsToMany = BelongsToManyTransform.belongsToMany.reverse(
                input.belongsToMany,
              );
            } else if (f === 'using') {
              result.using = BelongsToManyTransform.using.reverse(input.using);
            } else if (f === 'fields') {
              result.fields = BelongsToManyTransform.fields.reverse(
                input.fields,
              );
            } else {
              result[f] = core[f];
            }
          }
        }
      }
    }
    return result;
  }

  constructor(init?: Partial<IBelongsToManyInit>, context?: IFieldContext) {
    super(init, context);
    this.store = new BelongsToManyStorage(this.transform(init));
  }
}
