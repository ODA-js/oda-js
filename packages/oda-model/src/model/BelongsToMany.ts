import { Map, Record } from 'immutable';

import { IBelongsToMany, IBelongsToManyInit, IBelongsToManyStore } from '../interfaces/IBelongsToMany';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField, IFieldInit } from '../interfaces/IField';
import { EntityRef } from './EntityRef';
import { Field } from './Field';
import { Relation } from './Relation';


// tslint:disable-next-line:variable-name
export const DefaultBelongsToMany: IBelongsToManyStore = {
  name: null,
  title: null,
  description: null,
  fields: null,
  opposite: null,
  belongsToMany: null,
  using: null,
  //storage
  single: false,
  stored: false,
  embedded: false,
  // name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const BelongsToManyTransform = {
  belongsToMany: {
    transform: (inp) => {
      if (inp) {
        return new EntityRef(inp);
      } else {
        return null;
      }
    },
    reverse: (inp) => {
      if (inp) {
        return inp.toString();
      } else {
        return null;
      }
    },
  },
  fields: {
    transform: (input: IFieldInit[]) => {
      if (input) {
        return Map<string, IField>(input.map(p => [p.name, new Field(p)]) as [string, IField][]);
      } else {
        return null;
      }
    },
    reverse: (input: Map<string, IField>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(i => i.toJS());
      } else {
        return null;
      }
    },
  },
  using: {
    transform: (inp) => {
      if (inp) {
        return new EntityRef(inp);
      } else {
        return null;
      }
    },
    reverse: (inp) => {
      if (inp) {
        return inp.toString();
      } else {
        return null;
      }
    },
  },
};

// tslint:disable-next-line:variable-name
export const BelongsToManyStorage = Record(DefaultBelongsToMany);

export class BelongsToMany
  extends Relation<IBelongsToManyInit, IBelongsToManyStore> implements IBelongsToMany {
  public get verb(): 'BelongsToMany' {
    return 'BelongsToMany';
  }
  public get ref(): IEntityRef {
    return this.store.get('belongsToMany', null);
  }
  public get belongsToMany(): IEntityRef {
    return this.store.get('belongsToMany', null);
  }
  public get using(): IEntityRef {
    return this.store.get('using', null);
  }

  protected transform(input: Partial<IBelongsToManyInit>): Partial<IBelongsToManyStore> {
    const result: IBelongsToManyStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'belongsToMany') {
            result.belongsToMany = BelongsToManyTransform.belongsToMany.transform(input.belongsToMany);
          } else if (f === 'using') {
            result.using = BelongsToManyTransform.using.transform(input.using);
          } else if (f === 'fields') {
            result.fields = BelongsToManyTransform.fields.transform(input.fields);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IBelongsToManyStore> & Readonly<IBelongsToManyStore>): Partial<IBelongsToManyInit> {
    const result: IBelongsToManyInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'belongsToMany' || f === 'ref') {
            result.belongsToMany = BelongsToManyTransform.belongsToMany.reverse(input.belongsToMany);
          } else if (f === 'using') {
            result.using = BelongsToManyTransform.using.reverse(input.using);
          } else if (f === 'fields') {
            result.fields = BelongsToManyTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IBelongsToManyInit> = {}) {
    super();
    this.store = new BelongsToManyStorage(this.transform(init));
    this.init = new (Record<Partial<IBelongsToManyInit>>(init))();
  }
}
