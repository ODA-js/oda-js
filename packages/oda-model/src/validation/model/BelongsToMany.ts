import { IRelationPropsStore } from '../interfaces/IRelation';
import { Record } from 'immutable';
import { Map, Set } from 'immutable';

import { Persistent } from './Persistent';
import { transformMap, transformSet } from './utils';
import {
  IBelongsToManyRelationPropsStore,
  IBelongsToManyRelationProps,
  IBelongsToManyRelation,
} from '../interfaces/IBelongsToManyRelation';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField } from '../interfaces/IField';

// tslint:disable-next-line:variable-name
export const DefaultBelongsToMany: IBelongsToManyRelationPropsStore = {
  modelType: 'relation',
  name: null,
  title: null,
  description: null,
  verb: null,
  ref: null,
  fields: null,
  opposite: null,
  belongsToMany: null,
  //storage
  single: true,
  stored: true,
  embedded: true,
// name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const RelationTransform: {[k in keyof IBelongsToManyRelationPropsStore]?: any } = {
  fields: transformMap<IField>(),
};

// tslint:disable-next-line:variable-name
export const BelongsToManyStorage = Record(DefaultBelongsToMany);

export class BelongsToMany
  extends Persistent<IBelongsToManyRelationProps, IBelongsToManyRelationPropsStore> implements IBelongsToManyRelation {
  public get modelType(): 'relation' {
    return 'relation';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get ref(): IEntityRef {
    return this.store.get('ref', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get opposite(): string {
    return this.store.get('opposite', null);
  }
  public get single(): boolean {
    return this.store.get('single', null);
  }
  public get stored(): boolean {
    return this.store.get('stored', null);
  }
  public get embedded(): boolean {
    return this.store.get('embedded', null);
  }
  public get fullName(): string {
    return this.store.get('fullName', null);
  }
  public get normalName(): string {
    return this.store.get('normalName', null);
  }
  public get shortName(): string {
    return this.store.get('shortName', null);
  }
  public get verb(): 'BelongsToMany' {
    return 'BelongsToMany';
  }
  public get belongsToMany(): IEntityRef {
    return this.store.get('belongsToMany', null);
  }

  protected transform(input: IBelongsToManyRelationProps): IBelongsToManyRelationPropsStore {
    return {
      ...input,
      single: true,
      stored: true,
      embedded: true,
      verb: 'BelongsToMany',
      fields: RelationTransform.fields.transform(input.fields),
    };
  }
  protected reverse(input: IBelongsToManyRelationPropsStore): IBelongsToManyRelationProps {
    return {
      name: input.name,
      title: input.title,
      description: input.description,
      modelType: input.modelType,
      fullName: input.fullName,
      normalName: input.normalName,
      shortName: input.shortName,
      opposite: input.opposite,
      ref: input.ref,
      verb: input.verb,
      belongsToMany: input.belongsToMany,
      fields: RelationTransform.fields.reverse(input.fields),
    };
  }
  constructor(init: IBelongsToManyRelationProps) {
    super();
    this.store = new BelongsToManyStorage(this.transform(init));
    this.init = new (Record<IBelongsToManyRelationProps>(init))();
  }
}
