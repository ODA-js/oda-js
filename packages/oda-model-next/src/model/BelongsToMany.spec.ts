import 'jest';
import { BelongsToMany } from './BelongsToMany';

describe('BelongsToMany', () => {
  let relation: BelongsToMany;

  beforeAll(() => {
    expect(() => new BelongsToMany()).not.toThrow();
  });

  beforeEach(() => {
    relation = new BelongsToMany();
  });

  it('create empty', () => {
    expect(relation).toMatchSnapshot();
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('updates strings', () => {
    expect(() =>
      relation.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
      }),
    ).not.toThrow();
    expect(relation.name).toBe('cool');
    expect(relation.description).toBe('very cool');
    expect(relation.title).toBe('very cool title');
    expect(() =>
      relation.updateWith({
        name: 'cool!',
      }),
    ).not.toThrow();
    expect(relation.name).toBe('cool!');
  });

  it('updates with null or undefined', () => {
    expect(() =>
      relation.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
      }),
    ).not.toThrow();
    expect(relation.toJS()).toMatchSnapshot();
    expect(() =>
      relation.updateWith({
        description: null,
        title: undefined,
      }),
    ).not.toThrow();
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('toJS with dupes', () => {
    expect(() =>
      relation.updateWith({
        fields: [{ name: 'one' }, { name: 'one' }],
      }),
    ).not.toThrow();

    expect(relation.ref).toBeNull();
    expect(relation.fields.size).toBe(1);
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('toJS with HashMap', () => {
    expect(() =>
      relation.updateWith({
        fields: {
          one: {},
        },
      }),
    ).not.toThrow();

    expect(relation.ref).toBeNull();
    expect(relation.fields.size).toBe(1);
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('toJS ', () => {
    expect(() =>
      relation.updateWith({
        belongsToMany: 'id@Some#id',
        using: 'id@SomeOther#id',
      }),
    ).not.toThrow();

    expect(relation.ref.toJS()).toMatchSnapshot();
    expect(relation.using.toJS()).toMatchSnapshot();
    expect(relation.toJS()).toMatchSnapshot();
  });
});
