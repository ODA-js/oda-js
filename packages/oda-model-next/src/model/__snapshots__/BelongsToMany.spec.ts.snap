// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BelongsToMany create empty 1`] = `
BelongsToMany {
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "fields": null,
    "opposite": null,
    "belongsToMany": null,
    "using": null,
    "single": false,
    "stored": false,
    "embedded": false,
    "fullName": null,
    "normalName": null,
    "shortName": null,
  },
}
`;

exports[`BelongsToMany create empty 2`] = `
Object {
  "embedded": false,
  "single": false,
  "stored": false,
}
`;

exports[`BelongsToMany toJS  1`] = `
Object {
  "backField": "id",
  "entity": "Some",
  "field": "id",
}
`;

exports[`BelongsToMany toJS  2`] = `
Object {
  "backField": "id",
  "entity": "SomeOther",
  "field": "id",
}
`;

exports[`BelongsToMany toJS  3`] = `
Object {
  "belongsToMany": "id@Some#id",
  "embedded": false,
  "single": false,
  "stored": false,
  "using": "id@SomeOther#id",
}
`;

exports[`BelongsToMany toJS with HashMap 1`] = `
Object {
  "embedded": false,
  "fields": Array [
    Object {
      "name": "one",
    },
  ],
  "single": false,
  "stored": false,
}
`;

exports[`BelongsToMany toJS with dupes 1`] = `
Object {
  "embedded": false,
  "fields": Array [
    Object {
      "name": "one",
    },
  ],
  "single": false,
  "stored": false,
}
`;

exports[`BelongsToMany updates with null or undefined 1`] = `
Object {
  "description": "very cool",
  "embedded": false,
  "name": "cool",
  "single": false,
  "stored": false,
  "title": "very cool title",
}
`;

exports[`BelongsToMany updates with null or undefined 2`] = `
Object {
  "embedded": false,
  "name": "cool",
  "single": false,
  "stored": false,
}
`;
